File: ArithmeticSlices.py
# -*- coding: utf-8 -*-
"""
Created on Fri Mar 20 00:31:02 2020

@author: WELCOME
"""
"""
Arithmetic Slices
DP
Time - O(N)
Space - O(N)
"""

class Solution:
    def numberOfArithmeticSlices(self, A: List[int]) -> int:
        sum=0
        arr=[0]*len(A)
        for i in range(2,len(A)):
            if A[i]-A[i-1]==A[i-1]-A[i-2]:
                arr[i]+=arr[i-1]+1
                sum+=arr[i]
        return sum

File: ArithmeticSlices.java
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int numberOfArithmeticSlices(int[] A) {
        if(A == null || A.length == 0){
            return 0;
        }
        int count = 0;
        int[] dp = new int[A.length];
        
        for(int i = dp.length - 3; i >= 0; i--){
            if(A[i+1] - A[i] == A[i+2] - A[i+1]){
                dp[i] = 1 + dp[i+1];
                count += dp[i];
            }
        }
        
        return count;
    }
}